File: frontend/src/stores/visualizer.ts
import { ref } from 'vue'
import { defineStore } from 'pinia'

import type {
  ColorScheme,
  HydraPaletteMode,
  HydraQuality,
  VisualizerOptionsResponse,
  VisualizerPreset,
  VisualizerState,
} from '../services/api'
import {
  fetchVisualizerOptions,
  fetchVisualizerState,
  updateVisualizerState,
} from '../services/api'

export const useVisualizerStore = defineStore('visualizer', () => {
  const activePreset = ref<VisualizerPreset>('warehouse')
  const intensity = ref(65)
  const speed = ref(55)
  const brightness = ref(70)
  const colorScheme = ref<ColorScheme>('acid')
  const hydraColorfulness = ref(78)
  const hydraSceneChangeRate = ref(46)
  const hydraSymmetryAmount = ref(38)
  const hydraFeedbackAmount = ref(24)
  const hydraQuality = ref<HydraQuality>('medium')
  const hydraAudioReactivityEnabled = ref(true)
  const hydraPaletteMode = ref<HydraPaletteMode>('auto')
  const overlayMode = ref<VisualizerState['overlay_mode']>('logo')
  const autoCycleEnabled = ref(false)
  const autoCycleIntervalSeconds = ref(600)
  const updatedAt = ref<string | null>(null)
  const presets = ref<VisualizerOptionsResponse['presets']>([])
  const colorSchemes = ref<VisualizerOptionsResponse['color_schemes']>([])
  const hydraQualities = ref<VisualizerOptionsResponse['hydra_qualities']>([])
  const hydraPaletteModes = ref<VisualizerOptionsResponse['hydra_palette_modes']>([])

  async function refresh() {
    applyState(await fetchVisualizerState())
  }

  async function refreshOptions() {
    const options = await fetchVisualizerOptions()
    presets.value = options.presets
    colorSchemes.value = options.color_schemes
    hydraQualities.value = options.hydra_qualities
    hydraPaletteModes.value = options.hydra_palette_modes
  }

  async function save(nextState: Omit<VisualizerState, 'updated_at'>) {
    const token = localStorage.getItem('raveberg-admin-token')
    if (!token) {
      throw new Error('Admin session required')
    }

    applyState(await updateVisualizerState(nextState, token))
  }

  function applyState(state: VisualizerState) {
    activePreset.value = state.active_preset
    intensity.value = state.intensity
    speed.value = state.speed
    brightness.value = state.brightness
    colorScheme.value = state.color_scheme
    hydraColorfulness.value = state.hydra_colorfulness
    hydraSceneChangeRate.value = state.hydra_scene_change_rate
    hydraSymmetryAmount.value = state.hydra_symmetry_amount
    hydraFeedbackAmount.value = state.hydra_feedback_amount
    hydraQuality.value = state.hydra_quality
    hydraAudioReactivityEnabled.value = state.hydra_audio_reactivity_enabled
    hydraPaletteMode.value = state.hydra_palette_mode
    overlayMode.value = state.overlay_mode
    autoCycleEnabled.value = state.auto_cycle_enabled
    autoCycleIntervalSeconds.value = state.auto_cycle_interval_seconds
    updatedAt.value = state.updated_at
  }

  return {
    activePreset,
    intensity,
    speed,
    brightness,
    colorScheme,
    hydraColorfulness,
    hydraSceneChangeRate,
    hydraSymmetryAmount,
    hydraFeedbackAmount,
    hydraQuality,
    hydraAudioReactivityEnabled,
    hydraPaletteMode,
    overlayMode,
    autoCycleEnabled,
    autoCycleIntervalSeconds,
    updatedAt,
    presets,
    colorSchemes,
    hydraQualities,
    hydraPaletteModes,
    refresh,
    refreshOptions,
    save,
    applyState,
  }
})

File: backend/app/schemas/visualizer.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


VisualizerPreset = Literal[
    "particles",
    "kaleidoscope",
    "warehouse",
    "storm_lightning",
    "retro_cube",
    "retro_pipes",
    "dvd_bounce",
    "matrix_screen",
    "nebel",
    "vanta_halo",
    "hydra_rave",
    "hydra_chromaflow",
]
ColorScheme = Literal["mono", "acid", "ultraviolet", "redline"]
OverlayMode = Literal["logo", "qr", "off"]
HydraQuality = Literal["low", "medium", "high"]
HydraPaletteMode = Literal["auto", "neon", "warm", "cold", "acid"]


class VisualizerStateRead(BaseModel):
    active_preset: VisualizerPreset
    intensity: int
    speed: int
    brightness: int
    color_scheme: ColorScheme
    hydra_colorfulness: int
    hydra_scene_change_rate: int
    hydra_symmetry_amount: int
    hydra_feedback_amount: int
    hydra_quality: HydraQuality
    hydra_audio_reactivity_enabled: bool
    hydra_palette_mode: HydraPaletteMode
    overlay_mode: OverlayMode
    auto_cycle_enabled: bool
    auto_cycle_interval_seconds: int
    updated_at: datetime | None


class VisualizerStateUpdate(BaseModel):
    active_preset: VisualizerPreset
    intensity: int = Field(ge=0, le=100)
    speed: int = Field(ge=0, le=100)
    brightness: int = Field(ge=0, le=100)
    color_scheme: ColorScheme
    hydra_colorfulness: int = Field(ge=0, le=100)
    hydra_scene_change_rate: int = Field(ge=0, le=100)
    hydra_symmetry_amount: int = Field(ge=0, le=100)
    hydra_feedback_amount: int = Field(ge=0, le=100)
    hydra_quality: HydraQuality
    hydra_audio_reactivity_enabled: bool
    hydra_palette_mode: HydraPaletteMode
    overlay_mode: OverlayMode
    auto_cycle_enabled: bool
    auto_cycle_interval_seconds: int = Field(ge=300, le=1800)


class VisualizerOptionsResponse(BaseModel):
    presets: list[VisualizerPreset]
    color_schemes: list[ColorScheme]
    hydra_qualities: list[HydraQuality]
    hydra_palette_modes: list[HydraPaletteMode]


class VisualizerPresetOrderRead(BaseModel):
    presets: list[VisualizerPreset]
    skipped_presets: list[VisualizerPreset]


class VisualizerPresetOrderUpdate(BaseModel):
    presets: list[VisualizerPreset] = Field(min_length=1)
    skipped_presets: list[VisualizerPreset] = Field(default_factory=list)
